ECS damageable

ElectroJr opened the pull request:
# About the PR

ECS-ifies DamageableComponent and adds the DamageData class. ~~Time to start a sequel trilogy!~~

Currently marked as a draft because I still need re-review my own code and test that all the different damage effects still work properly after the damageable refactor trilogy shenanigans.

Even after these changes, DamageableComponent will still has a little bit of logic in it, for handling the networked component state and the radiation & explosion behaviours. I left radiation & explosions alone as others are currently working on re-doing radiation, and explosions are being modified in #4472.

# Changes

## Overview of changed files

The files with the most significant changes are:
- Almost all logic in `DamageableComponent` was moved into the newly created `DamageableSystem`
- `DamageData` was added
- `ResistanceSetPrototype` was changed around a bit to work with `DamageData`
- Damage related .cs files were relocated & name-spaces changed. In particular, all prototypes were moved into a prototype folder & namespace. Somewhere along the way github diff got confused and some files now show up as deleted & created instead of moved & modified.

As part of the ECSification, several bare-bones systems were created in order to subscribe to damage changed events. Components with new or modified systems include:
- Destructible
- Repairable
- DoAfter
- PoweredLight
- Window
- MobState

Many other files have small changes:
- Damage data-fields now use DamageData, leading to small changes in many .cs & .yml files
- DamageContainerPrototype & ResistanceSetPrototype ID's were changed, leading to more small changes in .yml files.
- Removing IDamageableComponent & changing damage namespaces also lead to lots of small changes.

Relevant content & integration tests were also added or modified.

## DamageData - What & Why?

Currently there are a couple of issues with damage:
- Logic in DamageableComponent is unnecessarily complicated (my own fault)
- Adding damage data-fields to components without prototype references is a pain (espc w/o prototype references)
- Damage changing component effects are not specified consistently in yaml files
  - Some components use two fields: a prototype string id & an integer
  - Some require dictionaries of strings & integers
  - Some expect damage type ID strings, others expected damage group IDs

The DamageData is a new class that tries to address these these issues. It has the `[DataDefinition]` attribute and is supposed to be used for yaml data-fields. It has all the logic for dividing damage groups into types and the application of resistance sets. The damageable component logic now only cares about damage types, not groups, which makes it a lot less convoluted. Additionally, the damage data-fields are now simply and consistently defined in c# using:
```cs
[ViewVariables(VVAccess.ReadWrite)]
[DataField("damage", required:true)] public DamageData Damage = default!;
```
Whereas previously you needed something like: 
```cs
[ViewVariables(VVAccess.ReadWrite)]
[DataField("damage")] public int Damage = 5;

[DataField("damageType"), required: true]
private readonly string _damageTypeID = "";

[ViewVariables(VVAccess.ReadWrite)]
public DamageTypePrototype DamageType = default!;

protected override void Initialize()
{
    base.Initialize();
    DamageType = IoCManager.Resolve<IPrototypeManager>().Index<DamageTypePrototype>(_damageTypeID);
}
```

The DamageData class also provides basic maths operators, which make it easier when scaling collections of damage types (for example, see the current code for scaling laser damage in  `ServerBatteryBarrelComponent`).

### But why is it in the ECSification PR?

Could this change have been made separately from ECS-ification? Yes, but I really can't be bothered ECS-ing damageable, only to end up deleting all the changes once again in a separate DamageData refactor. I just want to stop refactoring damage and get it over with. If it is prefered that they be separated (or that DamageData just be removed altogether), I can do that.

## Other changes

### Damage adjustments
While changing yaml fields to consistently use DamageData, there were a lot of oddities. For example, the combat knife (and many similar items) were just dealing the default damage type (blunt). No wonder surgery doesn't work, all the scalpels might as well have been bricks!

Additionally, healing damage is now "wasteful". Previously healing 30 brute on someone with 25/5/0 blunt/slash/piercing damage would heal them down to 0/0/0. Now it will only heal to 15/0/0 (10 healing for each damage type). This is a consequence of moving the damage-group splitting logic out of `DamageableComponent` and into `DamageData`, but arguably has the benefit that it encourages the use of more specialized/selective medicines when applicable. This change is usually not noticeable, but there are some exceptions like omnizine, which had to have it's healing amount scaled slightly to work as expected (not needed once damage units are added).


### Renaming prototype IDs
Another change that was made after some discord discussions (in howdoicode), was to rename some damage related prototype IDs. Currently some prototypes have redundant suffixes, e.g. the `metallicDamageContainer` DamageContainerPrototype. So yaml lines like:
```yaml
- type: Damageable
  damageContainer: metallicDamageContainer
  resistances: glassResistances
```
now look like:
```yaml
- type: Damageable
  damageContainer: Inorganic
  resistanceSet: Glass
```

### Destructible tests

Previously destructible components had a `DestructibleThresholdReachedMessage`, that was used exclusively for integration tests via the test-exclusive `TestThresholdListenerComponent`. I've just removed those, and instead created a `DestructibleThresholdListenerSystem` in `Content.IntegrationTests.Tests.Destructible`.

### Guns firing abstract bullets make for ineffective weapons

After discussion on discord in howdoicode, I've made bullets non-abstract entities.

Previously, bullets were marked as abstract entities, despite being "real" entities spawned in by guns. As a result, they do not get deserialization hooks called on them. This was fine before, but is a problem when using DamageData data-fields. Bullets were probably made abstract so that they do not appear in the sandbox spawn menu. However, as discussed on discord, this should probably be made a separate "hidden: true" property to distinguish them from "truly abstract" entities.

**Changelog**
:cl:
- tweak: Many components have had damage values & types adjusted. E.g., knives are no longer dull and actually deal slash damage.
- tweak: Eggs, tomatoes, and beakers are now destroyed on impact when thrown, instead of only when they land.

ElectroJr commented on Content.Server/Body/Respiratory/RespiratorComponent.cs:
```diff
@@ -358,27 +347,19 @@ private void TakeSuffocationDamage()
-            }
-
-            damageable.TryChangeDamage(DamageType, _damage, false);
+            Owner.EntityManager.EventBus.RaiseLocalEvent(Owner.Uid, new TryChangeDamageEvent(Damage, true), false);
```
In general I've removed 'TryGetComponent' or 'HasComponent' tests in favour of just raising TryChangeDamage events (which just do nothing if there is no component). But maybe events are computationally expensive and checking for applicable components beforehand is proffered?

mirrorcult replied:
they're not much different perf wise, but should get another maintainers opinion as to whether this should be a damageable system public method or an event; i'm leaning towards a public method, since its unlikely anything else will need to subscribe to this event (damageable can raise its own event for receiving feedback)

ElectroJr commented on Content.Shared/Damage/DamageableSystem.cs:
```diff
@@ -0,0 +1,291 @@
+        /// </summary>
+        public bool IgnoreResistances { get;  }
+
+        /// <summary>
+        /// Output. Did the damage actually change? Maybe it was all resisted, or you healed someone at full health.
+        /// </summary>
+        public bool DidDamageChange { get; set; } = false;
```
Currently, there are very few bits of code (ignoring tests) that actually care about whether any damage was changed by the event. Maybe this should just be removed?

mirrorcult commented on Content.Client/Body/UI/BodyScannerDisplay.cs:
```diff
@@ -2,7 +2,7 @@
 using Content.Shared.Body.Mechanism;
 using Content.Shared.Body.Part;
-using Content.Shared.Damage.Components;
+using Content.Shared.Damage;
```
correct

mirrorcult commented on Content.Shared/Damage/DamageData.cs:
```diff
@@ -0,0 +1,286 @@
+    ///     Supports basic math operations to modify damage.
+    /// </remarks>
+    [DataDefinition]
+    public class DamageData : ISerializationHooks
```
i'm partial to DamageSpecifier, but very subjective, should ask on discord

mirrorcult commented on Content.Server/Destructible/DestructibleSystem.cs:
```diff
@@ -13,5 +15,46 @@ public class DestructibleSystem : EntitySystem
+        /// <summary>
+        /// Check if any thresholds were reached. if they were, execute them.
+        /// </summary>
+        public void CheckAndExectuteThresholds(EntityUid uid, DestructibleComponent component, DamageChangedEvent args)
```
Execute

ElectroJr replied:
- Renamed function to `Execute`.
- Renamed `Threshold` class to `DamageThreshold` to avoid confusion with other thresholds (e.g., hunger, thirst).
- Renamed `DestructibleThresholdReachedEvent` to `DamageThesholdReachedEvent` for consistency.

ElectroJr commented:
After some discord discussion I've updated the PR, with some significantly changes to the code in order to:
- Remove prototype indexing in `ISerializationHooks.AfterDeserialization()`.
- Use Prototype IDs instead of prototypes for lists or dictionary keys, and only resolve prototypes when necessary.
- Add custom type serializers to all the prototype ID data fields.

ElectroJr commented on Content.Shared/Damage/DamageSpecifier.cs:
```diff
@@ -0,0 +1,257 @@
+        /// <summary>
+        ///     Main DamageSpecifier dictionary. Most DamageSpecifier functions exist to somehow modifying this.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public Dictionary<string, int> DamageDict
+        {
+            // Cannot use ISerializationHooks.AfterDeserialization(). So instead using this to index damage
+            // prototypes after deserialization only when actually needed.
+            get => GetCombinedDamageDictionary();
+            set => _damageDict = value;
+        }
+        private Dictionary<string, int>? _damageDict;
```
Previously I just used `ISerializationHooks.AfterDeserialization()` to initialise  `DamageDict`. The initialisation basically involves taking damage values specified by groups and dividing them into individual damage types. Without this, or some other initialisation function that somehow gets called after all prototypes are loaded in, I'm not sure how best to go about initialising `DamageDict`.

For now, I'm using this caching solution, but hopefully there's a better way to do it. Here, the `GetCombinedDamageDictionary()` function either returns `_damageDict` if it is not null, otherwise it runs through the initialisation steps that were previously in ` ISerializationHooks.AfterDeserialization()`

ElectroJr commented on Content.Server/Chemistry/ReagentEffects/HealthChange.cs:
```diff
@@ -17,53 +13,19 @@ namespace Content.Server.Chemistry.ReagentEffects
 
         /// <summary>
-        /// Damage group to change.
+        /// Damage to apply when 1u of reagent is consumed.
```
Ah yep, thanks for spotting that. Updated to:
> Damage to apply every metabolism cycle.

Also updated the HealthChange class doc from:
> Default metabolism for medicine reagents. Attempts to find a DamageableComponent on the target, and to update its damage values.

To:
> Default metabolism for medicine reagents. Raises a \<see cref="TryChangeDamageEvent"/\> every metabolism cycle while this reagent is in the bloodstream.

ElectroJr commented on Content.Shared/Damage/DamageableSystem.cs:
```diff
@@ -0,0 +1,313 @@
+        /// <returns>
+        ///     False if the given type is not supported or no damage change occurred; true otherwise.
+        /// </returns>
+        public static bool TryChangeDamage(DamageableComponent component, string damageType, int changeAmount)
```
OK, I've changed that function and functions in `PoweredLightSystem`, `WindowSystem` and `SharedMobStateSystem` to be non-static. I think those were all the instances where I did that.

ElectroJr commented on Content.Shared/Damage/DamageableSystem.cs:
```diff
@@ -0,0 +1,313 @@
+            if (component.ResistanceSetID != null)
+            {
+                _prototypeManager.TryIndex(component.ResistanceSetID, out component.ResistanceSet);
+            }
```
fixed

ElectroJr commented on Content.Shared/Damage/DamageableComponent.cs:
```diff
@@ -0,0 +1,126 @@
+        public DamageableComponentState(
+            Dictionary<string, int> damagePerType,
+            Dictionary<string, int>  damagePerGroup,
+            ResistanceSetPrototype? resistanceSet) 
```
fixed

ElectroJr commented on Content.Shared/Damage/DamageableComponent.cs:
```diff
@@ -0,0 +1,126 @@
+        public string? ResistanceSetID;
+
+        [ViewVariables(VVAccess.ReadWrite)]
+        public ResistanceSetPrototype? ResistanceSet;
```
fixed

ElectroJr commented on Content.Shared/Damage/DamageableSystem.cs:
```diff
@@ -0,0 +1,313 @@
+        /// <returns>
+        ///     Returns false if a no damage change occurred; true otherwise.
+        /// </returns>
+        public void TryChangeDamage(EntityUid uid, DamageableComponent component, TryChangeDamageEvent args)
```
The event has been removed now and replaced with a public method that takes damage and an entity as an argument. Currently the system just handles the damage itself if the entity has a damageable component. 

I made it take an entity as an argument based on sloth's comment. Passing the entity is useful if future components/systems have their own damage handling functionality (e.g., body system). That way you dont need to refactor every single damage dealing/healing system, and only need to add a bit of code to damageable system so that it can pass it on to the appropriate system.

ElectroJr commented on Content.Shared/Damage/DamageableSystem.cs:
```diff
@@ -0,0 +1,313 @@
+                }
+            }
+
+            DamageChanged(uid, component, false);
```
Currently initialising with damage isn't supported, so yeah it should just be removed (which I have now done). However making this change had a sort of domino effect that required several other changes that people may want to look at. So here is the reasoning behind those:

This dirty-on-init behaviour was around before even the damageable refactor and as far as I can tell the MobStateComponent appears to have inadvertently become reliant on the damage-changed event/message to effectively do some initialisation for it? So I've now also made changes to `MobStateComponent` to fix that.

AFAICT these changes accidentally fixed an error in `DisposalUnitTest` where the test was encountering an error when spawning entities (specifically the human dummy with a `MobStateComponent`). Because this was done in an un-awaited async function the errors were never raised and the test appeared to be succeeding? My changes in `MobStateComponent` fixed that error, so the test was actually running and getting to the actuall test assertions, which were failing. So I've also made changes to the disposal unit test so that it actually succeeds now.

ElectroJr commented on Content.Shared/Damage/DamageableSystem.cs:
```diff
@@ -0,0 +1,322 @@
+
+            component.DamagePerType = state.DamagePerType;
+
+            // Calculate dependent values and raise a local event. The event is needed as there may be client-exclusive
+            // systems (e.g. UI) that need to know if damage changed as a result of server-exclusive damage-dealing
+            // systems.
+            DamageChanged(component, component.TotalDamage < newTotalDamage);
```
So here, this is raising `DamageChangedEvents` inside of the handle damageable component state function.  This feels wrong, but seems like the most reasonable way to ensure that client-side systems are informed when server-side systems deal damage.

Currently this doesn't really matter. Things like MobState are `Dirty()`-ed by the server. But it might matter for some future client-exclusive systems?

ElectroJr commented:
Reviews have been addressed.

Other changes:
- Damage system events/methods should now all use Uids instead of IEntity.
- Due to a domino effect, addressing the comment about `Dirty()` calls during damageable component initialization required changes to MobState component and disposal unit tests.
  - MobState was relying on DamageChanged events to initialize itself.
  - DisposalUnitTest was broken/failing, but was not evident due to a null error in MobState???

ElectroJr commented:
Some more changes:
- DamageableComponent is now friends with DamageableSystem
- DamageableComponent can now be initialised with some damage, in case someone ever wants to create a pre-damaged entity prototype.
- DamageChangedEvent now also has the amount by which damage changed as an argument.
- DamageableComponent now uses a DamageSpecifier to store damage information
- DamageSpecifier was slightly extended.
  - It now supports `Clamp()` operations that cap the min/max damage. 
  - `DamageableSystem.GetDamagePerGroup()` was moved into `DamageSpecifier`

Zumorica reviewed (approved):
looks good to me
